PosteAPI/views.py

```diff
@@ -146,3 +150,192 @@ def post(self, request):
                 {"result": {"success": False, "errors": serializer.errors}},
                 status=status.HTTP_400_BAD_REQUEST,
             )
+
+
+class FolderAPI(APIView):
+    @swagger_auto_schema(
+        operation_description="Returns a list of all folders",
+        responses={
+            200: FolderSerializer(many=True),
+            400: "Bad Request",
+        },
+    )
+    def get(self, request):
+        folders = Folder.objects.all()
+        serializer = FolderSerializer(folders, many=True)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    @swagger_auto_schema(
+        operation_description="Creates a new folder.",
+        request_body=FolderCreateSerializer,
+        responses={
+            201: openapi.Response(
+                description="The created folder object.", schema=FolderSerializer
+            ),
+            400: openapi.Response(
+                description="Bad Request",
+                examples={
+                    "application/json": {
+                        "title": ["title cannot be blank"],
+                        "creator": ["creator is not a valid user"],
+                    }
+                },
+            ),
+        },
+    )
+    def post(self, request):
+        serializer = FolderCreateSerializer(data=request.data)
+        if serializer.is_valid():
+            folder = serializer.save()
+            response = Response(
+                FolderSerializer(folder).data, status=status.HTTP_201_CREATED
+            )
+            return response
+        else:
+            response = Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+            # error message contained in response.data
+            return response
+
+
+class FolderForUser(APIView):
+    def get_object(self, pk):
+        try:
+            return User.objects.get(pk=pk)
+        except User.DoesNotExist:
+            return None
+
+    def get_owned(self, user):
+        if Folder.objects.filter(creator=user).exists():
+            return Folder.objects.filter(creator=user)
+        else:
+            return None
+
+    def get_shared(self, user):
+        if FolderPermission.objects.filter(user=user).exists():
+            shared = None
+            for permission in FolderPermission.objects.filter(user=user):
+                if shared is None:
+                    shared = Folder.objects.filter(pk=permission.folder.pk)
+                else:
+                 shared = shared | Folder.objects.filter(pk=permission.folder.pk)
```

Comment:
indentation error here may / may not cause an issue, python is wonky with whitespace

Reply:
Fixed indentation.